### ISSUE TITLE
deterministic sorting for all_orm_descriptors

### ISSUE THREAD

**Issue by sevens-ef** [sqlalchemy/sqlalchemy#5494]
Not entirely sure if this should be a bug or a feature request, added it as the former.

**Describe the bug**
`hybrid_property`s don't have a `_creation_order` attribute, while most/all other `Column`s and properties from SQLAlchemy seem to have one (as far as I can tell from checking the codebase).

**Expected behavior**
`hybrid_property`s to have a `_creation_order` so it can be sorted predictably.

**To Reproduce**
<details>

```py
#!/usr/bin/env python
import sqlalchemy
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.ext.hybrid import hybrid_property

Base = declarative_base()


class Model(Base):
    __tablename__ = 'model'

    id = sqlalchemy.Column(sqlalchemy.Integer, primary_key=True)
    assert hasattr(id, '_creation_order')

    @hybrid_property
    def foo(self):
        return 'bar'

    assert not hasattr(foo, '_creation_order')  # Would expect this to fail


inspector = sqlalchemy.inspect(Model)
sorted_descriptors = sorted(list(inspector.all_orm_descriptors.items()),
                            key=lambda item: getattr(item[1], '_creation_order', 0))

# This fails (attribute 'foo' should be the second item in `sorted_descriptors`)
assert sorted_descriptors[1][0] == 'foo'
```

</details>

**Error**
N/a

**Versions.**
 - OS: Ubuntu 18.04
 - Python: 3.6.9
 - SQLAlchemy: 1.3.18
 - Database: PostgreSQL (but this shouldn't matter here ;) )
 - DBAPI: psycopg2

**Additional context**
We use this in our own code, including in a custom fork of ColanderAlchemy to deterministically sort the properties of a model.

Monkey-patching a `sqlalchemy.util.set_creation_order(self)` call to the end of `hybrid_property.__init__()` seems to fix this.

**Comment by zzzeek** [comment]
hey there -

I would classify this as "use case".

so firstly, there is currently no "creation order" setting on *any* Python descriptors in SQLAlchemy.   the creation order you see is actually configured on the Column objects and MapperProperty objects that are not actually the front-facing descriptors that your code accesses.   if you see a _creation_order that's because it is proxied outwards on the descriptor that houses a MapperProperty.

There are actually several descriptors in play, there's not just hybrids there are also association proxies.  anything that subclasses InspectionAttr.    I would prefer not to have a convention that every InspectionAttr subclass contains a set_creation_order() in its constructor, this implies another thing that can be forgotten, needs to be tested, etc.  I also would prefer not to use a metaclass here.

Then, the "_creation_order" attribute itself is a private attribute, and I would advise against releasing a third party library that is reading this attrbute directly.  A better request would be, "provide a means of iterating class-bound attributes in deterministic order".

From there, I would then refer to [pep 520](https://www.python.org/dev/peps/pep-0520/) which is in Python 3.6.   That is, you should be able to simply say dir(class), and that will give you all the descriptors in deterministic order IIUC what pep 520 is saying.

On the SQLAlchemy side, we have a method [mapper.all_orm_descriptors](https://docs.sqlalchemy.org/en/14/orm/mapping_api.html?highlight=all_orm_descriptors#sqlalchemy.orm.Mapper.all_orm_descriptors), which is what you would use to get a collection of all the descriptors that pertain to SQLAlchemy only.   This is a sorted list but I'm not fully certain that it is deterministically sorted or if it is guaranteed to take advantage of pep-520 guarantees.  If you could test this method for deterministic sorting under python 3.6 or greater, I will favor an approach that provides for this collection to be in pep-520 sorting.

**Comment by sevens-ef** [comment]
Hi, thanks for the response.

We're already using `Mapper.all_orm_descriptors` but are then doing the manual sorting on that :)

I just did some quick tests (Python 3.6.9), will look further into it (and can do further tests and investigations) next week after my weekend.  Results and findings are below.

At least the output of `Mapper.all_orm_descriptors.items()` doesn't seem to be deterministic.  I replaced the line creating `sorted_descriptors` in my reproduction code with `sorted_descriptors = inspector.all_orm_descriptors.items()` and the assert at the end fails ~half the time.

I don't know much about the `ImmutableProperties` class that `Mapper.all_orm_descriptors` uses, but does the sortabillity PEP 520 (hopefuly ;) ) provides trickle through to it?

Also tested with the following added to the reproduction code:

```!py
model_dir = dir(Model)
foo_index = model_dir.index('foo')
id_index = model_dir.index('id')

if foo_index > id_index:
    print('OK')
else:
    print('foo before id')
```

and this seems to always print `'foo before id'`, both in cases where the final `assert` fails as where it succeeds (so that part *is* deterministic, just in the wrong order?).  Further inspecting the `dir(Model)` output in a debug session confirms this (in all the 10-20 runs I tested).

Just as an extra check: the output of `print(inspector.all_orm_descriptors.keys().index('foo'))` also differs per run.

Also in a unrelated quicksearch through PEP 520 I found this in part 1 of the specification: "`dir()` will not depend on `__definition_order__`".

**Comment by zzzeek** [comment]
the collection is sorted but the way things get into it is not.  so we would need to adjust that logic inside the instrumentation system.  it wouldn't be too difficult.

**Comment by zzzeek** [comment]
i have a patch here but the tricky thing is knowing how `cls.__dict__` was populated in all cases.   assuming dictionaries are insert-order as they are in Python 3.6, I can't think of a situation where things will not be deterministic but it's hard to be sure (things like classical mappings, mixin mappings, inheritance mappings, etc).

i think it will be one of those things where you'll test it and it will work great but some years down the road we'll find weird edge cases that don't work.   it's the best I can do for now.

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **master** branch:

**base all_orm_descriptors ordering on cls.__dict__ + cls.__mro__** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2131

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **rel_1_3** branch:

**base all_orm_descriptors ordering on cls.__dict__ + cls.__mro__** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2132

**Comment by zzzeek** [comment]
OK so that is merged.  it shouldn't break anything, and it definitely makes things better.  however, you should test it on your end and see if you can make it break, i can make additional changes if so.

**Comment by zzzeek** [comment]
oh also dir(cls) is not the pep-520 sorting, cls.__dict__ is where the sorting is.  but the routine I have here doesn't necessraily use this ordering directly if there's an inheritance hierarchy in play.

**Comment by sevens-ef** [comment]
Well that's quick, thanks!

> i think it will be one of those things where you'll test it and it will work great but some years down the road we'll find weird edge cases that don't work.

Yeah, that's often the nature of these things.

The handling of keys existing in both a sub- and super-class seems logical to me (though personally this part was easier to get from the comment in [instrumentation.py](https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2131/2/lib/sqlalchemy/orm/instrumentation.py#166) than the documentation).

I tested the fix (rel_1_3 branch) with a few variations of the reproduction script (including using sub-classes), seems to work fine (100 runs for each variation).

I also applied the patch manually to the older SQLAlchemy version our code needs atm, changed our explicit sorting to use `all_orm_descriptors` verbatim, ran one of our unittests that depends on the ordering 100 times, all passed.  This has the fields defined on the super-class on the model (both classes have many attributes of various types, both from SQLAlchemy and regular Python).  Ran the same unittest against the super-class directly as well (100x), all passed.

Unfortunately we need a pre-1.3 version of SQLAlchemy atm for our code so further testing is tricky; will keep you posted if we run into issues once we've upgraded.

N.b.: I noticed though that https://docs.sqlalchemy.org/en/13/orm/mapping_api.html#sqlalchemy.orm.Mapper.all_orm_descriptors mentions this behavior is introduced in version 1.4, but that documentation is for 1.3 :)
